astropy/coordinates/earth.py

```diff
@@ -287,22 +291,18 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
         ``gd2gc`` is used.  See https://github.com/liberfa/erfa
         """
         ellipsoid = _check_ellipsoid(ellipsoid, default=cls._ellipsoid)
-        # We use Angle here since there is no need to wrap the longitude -
-        # gd2gc will just take cos/sin anyway.  And wrapping might fail
-        # on readonly input.
-        lon = Angle(lon, u.degree, copy=False)
+        # As wrapping fails on readonly input, we do so manually
+        lon = Angle(lon, u.degree, copy=False).wrap_at(180 * u.degree)
         lat = Latitude(lat, u.degree, copy=False)
         # don't convert to m by default, so we can use the height unit below.
         if not isinstance(height, u.Quantity):
             height = u.Quantity(height, u.m, copy=False)
         # get geocentric coordinates. Have to give one-dimensional array.
-        xyz = erfa.gd2gc(getattr(erfa, ellipsoid),
-                         lon.to_value(u.radian),
-                         lat.to_value(u.radian),
-                         height.to_value(u.m))
+        xyz = REPRESENTATION_CLASSES[ellipsoid.lower() + "geodetic"](
+            lon, lat, Distance(height, allow_negative=True), copy=False,
+        ).represent_as(CartesianRepresentation).get_xyz(xyz_axis=-1)
         self = xyz.ravel().view(cls._location_dtype,
```

Comment:
Might be clearer to change the unit of `xyz` here instead of `self` at the end - mostly as we can do it in-place.